src/include/migraphx/par.hpp

```diff
@@ -0,0 +1,133 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#ifndef MIGRAPHX_GUARD_MIGRAPHX_PAR_HPP
+#define MIGRAPHX_GUARD_MIGRAPHX_PAR_HPP
+
+#include <migraphx/config.hpp>
+#if MIGRAPHX_HAS_EXECUTORS
+#include <execution>
+#else
+#include <migraphx/simple_par_for.hpp>
+#endif
+#include <algorithm>
+#include <mutex>
+#include <vector>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+namespace detail {
+
+struct exception_list
+{
+    std::vector<std::exception_ptr> exceptions;
+    std::mutex m;
+    void add_exception()
+    {
+        std::lock_guard<std::mutex> guard(m);
+        exceptions.push_back(std::current_exception());
+    }
+    template <class F>
+    auto collect(F f)
+    {
+        return [f, this](auto&&... xs) {
+            try
+            {
+                f(std::forward<decltype(xs)>(xs)...);
+            }
+            catch(...)
+            {
+                this->add_exception();
+            }
+        };
+    }
+    void throw_if_exception() const
+    {
+        if(not exceptions.empty())
+            std::rethrow_exception(exceptions.front());
+    }
+};
+
+} // namespace detail
+
+template <class InputIt, class OutputIt, class UnaryOperation>
+OutputIt par_transform(InputIt first1, InputIt last1, OutputIt d_first, UnaryOperation unary_op)
+{
+#if MIGRAPHX_HAS_EXECUTORS
+    return std::transform(std::execution::par, first1, last1, d_first, std::move(unary_op));
+#else
+    simple_par_for(last1 - first1, [&](auto i) { d_first[i] = unary_op(first1[i]); });
+    return d_first + (last1 - first1);
+#endif
+}
+
+template <class InputIt1, class InputIt2, class OutputIt, class BinaryOperation>
+OutputIt par_transform(
+    InputIt1 first1, InputIt1 last1, InputIt2 first2, OutputIt d_first, BinaryOperation binary_op)
+{
+#if MIGRAPHX_HAS_EXECUTORS
+    return std::transform(
+        std::execution::par, first1, last1, first2, d_first, std::move(binary_op));
+#else
+    simple_par_for(last1 - first1, [&](auto i) { d_first[i] = binary_op(first1[i], first2[i]); });
+    return d_first + (last1 - first1);
+#endif
+}
+
+template <class InputIt, class UnaryFunction>
+void par_for_each(InputIt first, InputIt last, UnaryFunction f)
+{
+#if MIGRAPHX_HAS_EXECUTORS
+    // Propagate the exception
+    detail::exception_list ex;
+    std::for_each(std::execution::par, first, last, ex.collect(std::move(f)));
+    ex.throw_if_exception();
```

Comment:
Why is exception handled here at all, when the std::for_each should behave in a certain way, and this will alter that behavior. Thanks.

Reply:
It breaks some unit tests that tries to detect if an error occurs. I think its better to have an exception thrown which could possibly be recovered than to have `std::terminate` called. We could update the other algorithms the same way, but for now I updated this which fixes an unit tests.

Reply:
The logical way to recover from exceptions would be to handle each of those exceptions -- not just the first one, which isn't happening here.. ; However, as a philosophy, the parallel mode shouldn't be generating exceptions.. so which unit test is failing here specifically, please. Maybe that instance shouldn't be parallelized. Thanks.

Reply:
> The logical way to recover from exceptions would be to handle each of those exceptions -- not just the first one, which isn't happening here.. ;

We can only rethrow one exception at a time. We could wrap the exception but then we lose the backtrace when we rethrow. 

> However, as a philosophy, the parallel mode shouldn't be generating exceptions.

The only way to report errors from algorithms is to throw an exception. Checking for the errors before hand can be costly as well. Since we are calling many standard library functions which are `noexcept` then we need to recover from the errors for several reasons:

1) Unit tests can check that such errors are check for.
2) In the API functions we need to convert any exception to en error code, terminating the application is not a good option at all.

Furthermore, this is what is suggested by Anthony Williams the author of the  [C++ Concurrency in Action](https://www.manning.com/books/c-plus-plus-concurrency-in-action) book:

https://stackoverflow.com/a/55354021/375343

The C++ standard terminates the program since there is a lot of different ways to handle the multiple exceptions and not a clear best way. For now, I just throw the first exception. This should be good enough for our uses cases(unit test check the exception and the API returns an error code). 

> so which unit test is failing here specifically, please.

The gathernd check for out of bounds error here will fail:

https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/develop/test/ref/gathernd.cpp#L389

Which checking for this exception:

https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/develop/src/include/migraphx/op/gathernd.hpp#L198

Its too costly to move it out of the algorithm(as we would need to run a serial version and parallel version). 

> Maybe that instance shouldn't be parallelized.

Just because we check for out of bounds access and throw an exception doesnt mean we should not make it parallelized. And I see no compelling reason to have parallel algorithms not throw exceptions.

Reply:
Thanks for the detailed explanation :-) But I was referring just to the approach taken here: to parallelize std::transform() into par_for**, and wanted to understand the reason of test case failure. 

Shouldn't bounds checking be done outside of the parallelization logic..? Thanks.

Reply:
> But I was referring just to the approach taken here: to parallelize std::transform() into par_for**

I dont understand what you are referring to. `par_for` will propagate the exception. 

> Shouldn't bounds checking be done outside of the parallelization logic..?

That is too slow. We would need to practically do everything serially first to chekc the bounds and then run it in parallel which doesnt make sense.